File: python/search/LinearSearch.py
# LINEAR SEARCH

# This algorithm checks whether the given element is present in the array or not.
# If the element is found in the array it returns the index of that element else 
# it returns -1 if the element is not found in the array.

def LinearSearch(arr,key):
    # we use a for loop to iterate through the array. 
    for i in range(len(arr)):
        if arr[i]==key:
            return i   # this returns the index if the key is found
    
    return -1   # this value is returned if the key is not found in the array



# Starting with the main code
# Creating array of the size n
# Size would be taken into the variable n from user
def main():
    print("please enter the size of the array :")
    n=int(input()) # takes size as input
    array=[]
    print("please enter the elements of the array :")
    for j in range(n):     # for loop for taking the elements as input
        array.append(int(input()))
    print("please enter the key element :")    
    x=int(input())         # the key is taken as input
    print()
    ans=LinearSearch(array,x)    # linear search is called  and the result is stored
    if ans==-1:
        print("element is not found")
    else:
        print(f'the element is found at index {ans}')    


main()


""""
Example Test Cases:
1)

please enter the size of the array :
5
please enter the elements of the array :
1
2
3
4
5
please enter the key element :
4

the element is found at index 3

2)

please enter the size of the array :
5
please enter the elements of the array :
1
2
3
4
5
please enter the key element :
7

element is not found

Time Complexity  : O(n)
Space Complexity : O(1)
"""        

File: c++/search/linearSearch.cpp
/*
a linear search or sequential search is a method for finding an element within a list. It sequentially checks each element of the list until a match is found or the whole list has been searched.
linear search:-

(i) Start from the leftmost element of arr[] and one by one compare x with each element of arr[]
(ii) If x matches with an element, return the index.
(iii) If x doesn’t match with any of elements, return -1.
*/
#include <bits/stdc++.h>
using namespace std;

//Iterative Search: Time Complexity is O(n)
int linearSearchIterative(int n, int key, vector<int> v)
{
    //Loop will run util key is found in the vector or will return -1 if element is not there
    for (int i = 0; i < n; i++)
        if (v[i] == key)
            return i;
    return -1;
}

//Recursive Approach: Time Complexity is O(n)
int linearSearchRecursive(int i, int n, int key, vector<int> v)
{
    //If element not found then we will return -1
    if (i == n || i > n)
        return -1;
    //If element found then returning the position of the element
    if (v[i] == key)
        return i;
    //Calling the linearSearchRecursive function for i+1 if key is not found at ithe position util i become greater or equal to n
    return linearSearchRecursive(i + 1, n, key, v);
}

int main()
{
    int n, key, elem;

    //Taking the no of element i b/w we have to search a particular element
    cout << "Element the no of element: ";
    cin >> n;
    vector<int> v;

    cout << "Element the element of the array: ";
    for (int i = 0; i < n; i++)
    {
        cin >> elem;
        v.push_back(elem);
    }

    cout << "Enter the element which is to find: ";
    cin >> key;

    //This is for Iterative Approach for the linear Search
    cout << "---------- By Iterative Way ------------------------\n";
    if (linearSearchIterative(n, key, v) > 0)
        cout << "Element found at " << linearSearchIterative(n, key, v) + 1 << " Position.\n";
    else
        cout << "Element not found.\n";

    //And this is for Recursive Approach for the linear Search
    cout << "---------- By Recursive Way ------------------------\n";
    if (linearSearchRecursive(0, n, key, v) > 0)
        cout << "Element found at " << linearSearchRecursive(0, n, key, v) + 1 << " Position.\n";
    else
        cout << "Element not found.\n";
}
/*
Time Complexity: O(N)
Space complexity: O(1)
Sample I/O:
INPUT:
Enter number of elements in the array: 10
Element the element of the array: 3 323 232 434 32 23 233 26575 545 890
Enter the element which is to find: 233

OUTPUT:
Enter the element which is to find: 233
---------- By Iterative Way ------------------------
Element found at 7 Position.
---------- By Recursive Way ------------------------
Element found at 7 Position.
*/